### DISCUSSION TITLE
Is aggregate function available?

### DISCUSSION THREAD

**Discussion by TomasMoskala** [prisma/prisma#3438] [Q&A] [answered]
My PrismaClient does not have it
(https://github.com/prisma/prisma/releases/tag/2.2.0  Aggregations API)
```json
{
  "name": "typescript-graphql",
  "scripts": {
    "clean": "rm -rf dist",
    "build": "npm -s run clean && npm -s run generate && tsc",
    "generate": "npm -s run generate:prisma && npm -s run generate:nexus",
    "generate:prisma": "prisma generate",
    "generate:nexus": "ts-node --transpile-only src/schema",
    "postinstall": "npm -s run generate",
    "dev": "ts-node-dev --no-notify --respawn --transpileOnly src/server"
  },
  "dependencies": {
    "@apollo/client": "^3.1.3",
    "@nexus/schema": "0.13.1",
    "@prisma/cli": "^2.5.1",
    "@prisma/client": "^2.5.1",
    "@types/bcrypt": "^3.0.0",
    "@types/jsonwebtoken": "^8.5.0",
    "@types/node-fetch": "2.5.7",
    "@types/react": "^16.9.46",
    "apollo-client": "^2.6.10",
    "apollo-server": "^2.15.1",
    "bcrypt": "^5.0.0",
    "cross-fetch": "^3.0.5",
    "dotenv": "^8.2.0",
    "graphql": "14.6.0",
    "jsonwebtoken": "^8.5.1",
    "nexus-prisma": "^0.16.2",
    "react": "^16.13.1",
    "typescript": "^3.9.7"
  },
  "devDependencies": {
    "@types/node": "12.12.42",
    "ts-node": "8.10.1",
    "ts-node-dev": "1.0.0-pre.44"
  },
  "prettier": {
    "singleQuote": true,
    "semi": false,
    "trailingComma": "all"
  },
  "engines": {
    "node": ">=10.0.0"
  }
}

```

**Comment by nikolasburk** [comment, marked as answer]
Hey @TomasMoskala, thanks for raising this! Aggregations should be available by default (without a feature flag) if you're using a version greater than [v2.5.0](https://github.com/prisma/prisma/releases/tag/2.5.0) (which seems to be the case for you).

I was not able to reproduce the issue, I can use `aggregate` on version 2.5.1, e.g.:

```ts
const count = await prisma.user.aggregate({
  count: true
})
console.log(count)
```

If the problem persists, would you mind opening an issue so that our engineering team can debug it properly :) also, just as a side-note, make sure to delete `node_modules` and install `npm install @prisma/client` again, sometimes this helps too :)

**Reply by TomasMoskala** [reply]
Thanks! Somehow @prisma/client was not installed and everything worked so I did not notice.

I would like to ask about data protection if you don't mind. I expose wallet ballance at User objectType to include it in single query to save one server call. I guess that's what GraphQL was created for. At  the same time I need to show some User information to all other users, like name, surname etc. To protect sensitive information I wrote a resolver to the  t.model.wallet_balance to get the id from token

```ts
export const User = objectType({
  name: 'user',
  definition(t) {
    t.model.user_name()
    t.model.user_surname()
    t.model.wallet_balance({
      resolve: (_, __, ctx) => {
      const id_user = ctx.userToken.user_id && ctx.userToken.user_id > 0 ? ctx.userToken.user_id : 0; // get user id from token
      return ctx.prisma.wallet_balance.findMany({where: {id_user}})}
    })

```
This way any call will only return the current user's ballances. Works fine but is there a way to do this at the wallet_balance objectType or do I have to copy the code everywhere the table is connected?

**Reply by ryands17** [reply]
Hey @TomasMoskala 👋 
You could try something like this:

```ts
const User = objectType({
  name: 'User',
  definition(t) {
    t.model.id()
    t.model.name()
    t.model.surname()
    t.model.wallet_balance({
      async resolve(_parent, _args, ctx) {
        const user = await ctx.prisma.user.findOne({
          where: { id: 1 },
        })
        return user.wallet_balance
      },
    })
  },
})

const AnotherModel = objectType({
  name: 'AnotherModel',
  definition(t) {
    t.model.id()
    t.model('User').wallet_balance()
  },
})
```

**Reply by TomasMoskala** [reply]
wallet_balance is a table and it returns many (currencies) so I need to do ctx.prisma.wallet_balance.findMany.
But if I can still reference it t.model('User').wallet_balance() getting an array then it would be fine, Thanks.

**Reply by TomasMoskala** [reply]
Well I get an error TypeError: t.model is not a function. What is the right syntax?

**Reply by ryands17** [reply]
Try this one by creating a separate object type:

```ts
const WalletBalance = objectType({
  name: 'WalletBalance',
  definition(t) {
    t.model('User').wallet_balance( /* resolver here */)
  },
})

const User = objectType({
  name: 'User',
  definition(t) {
    t.model.id()
    t.model.name()
    t.model.surname()
    t.field('wallet_balance', {
      type: 'WalletBalance',
    })
  },
})
```

**Reply by TomasMoskala** [reply]
The new object works like this but I get an error for the user field - resolver missing

```
export const WalletBalance = objectType({
  name: 'WalletBalance',
  definition(t) {
    t.model('user').wallet_balance({
      resolve: (_, __, ctx) => {
      const id_user = ctx.userToken && ctx.userToken.user_id > 0 ? ctx.userToken.user_id : 0; // get user id from token
      return ctx.prisma.wallet_balance.findMany({where: {id_user}})}
    })
  },
})

```
error TS2345: Argument of type '{ type: "WalletBalance"; }' is not assignable to parameter of type 'NexusOutputFieldConfig<"user", "wallet_balance"> & { resolve: FieldResolver<"user", "wallet_balance">; }'.
  Property 'resolve' is missing in type '{ type: "WalletBalance"; }' but required in type '{ resolve: FieldResolver<"user", "wallet_balance">; }'.

36     t.field('wallet_balance', {
                                 ~
37       type: 'WalletBalance',
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
38     })
   ~~~~~

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
